### PR TITLE
Added keras util API for conversion of data tensor from channels_last to channels_first using MXNet backend

### PR DESCRIPTION
**Pull request by karan6181**
- A Keras Util API to convert image data format from channels_last to channels_first when using MXNet backend.
- Added unit test to test the functionality

@sandeep-krishnamurthy @roywei

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -4081,6 +4081,15 @@ def _calculate_conv_output_size(input_length, filter_size, padding, stride,
 @keras_mxnet_symbol
 def _preprocess_convnd_input(data_var, data_format):
     if data_format == 'channels_last' and ndim(data_var) > 3:
+        warnings.warn(
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: These are too many warnings. For example, if user adds, 1 Conv2D layer he gets same warning twice! one from preprocess_convnd and one from postprocess_convnd.

Please clean up these warning message. It should only once per operator.

File: keras/utils/np_utils.py [hunk 2]
```diff
@@ -48,3 +49,62 @@ def normalize(x, axis=-1, order=2):
     l2 = np.atleast_1d(np.linalg.norm(x, order, axis))
     l2[l2 == 0] = 1
     return x / np.expand_dims(l2, axis)
+
+def mxnet_image_channels_conversion(np_data):
```
Comment (sandeep-krishnamurthy) [inline keras/utils/np_utils.py]: * Why should we call it mxnet_image_channels_conversion?
API should be independent of a framework, it is just transposing an image. Please rename and make it generic.

* As we discussed earlier, it should take both single np_array and list of np_array. Think like a user, he loads a batch of data, and he will call this API to transpose.

May be below API?
```
def to_channels_first(data):
```

File: keras/utils/np_utils.py [hunk 3]
```diff
@@ -48,3 +49,62 @@ def normalize(x, axis=-1, order=2):
...
+
+    # Arguments
+        np_data: image data tensor
+
+    # Returns
+        A image data tensor with channels_first format
+    """
+    data_format = K.image_data_format()
+    if data_format not in {'channels_first', 'channels_last'}:
+        raise ValueError('MXNet Backend: Unknown data_format ' +
```
Comment (sandeep-krishnamurthy) [inline keras/utils/np_utils.py]: No MXNet things outside MXNet backend file.

File: keras/utils/np_utils.py [hunk 4]
```diff
@@ -48,3 +49,62 @@ def normalize(x, axis=-1, order=2):
...
+                         str(data_format))
+    if K.backend() == 'mxnet':
```
Comment (sandeep-krishnamurthy) [inline keras/utils/np_utils.py]: As discussed above, not required to restrict this to mxnet backend

File: keras/utils/np_utils.py [hunk 5]
```diff
@@ -48,3 +49,62 @@ def normalize(x, axis=-1, order=2):
...
+        if data_format == 'channels_first':
+            shape = np_data.shape
```
Comment (sandeep-krishnamurthy) [inline keras/utils/np_utils.py]: Handle both single and list of images.

File: keras/utils/np_utils.py [hunk 6]
```diff
@@ -48,3 +49,62 @@ def normalize(x, axis=-1, order=2):
...
+            if len(shape) == 5:
+                if shape[1] < shape[4]:
+                    warnings.warn('Image data tensor format is already '
+                                  '`channels_first`', stacklevel=2)
+                    return np_data
+                np_data = np.transpose(np_data, (0, 4, 1, 2, 3))
+            elif len(shape) == 4:
+                if shape[1] < shape[3]:
```
Comment (sandeep-krishnamurthy) [inline keras/utils/np_utils.py]: what if I have an image of 2*2 ?

File: tests/keras/utils/np_utils_test.py [hunk 7]
```diff
@@ -28,6 +29,38 @@ def test_to_categorical():
...
+    ]
+
+    expected_shapes = [
+        (10, 3, 32, 32),
+        (10, 3, 32, 32),
+        (10, 1, 24, 128, 128),
+        (10, 1, 24, 128, 128)
+    ]
+
+    inputs = [np.random.randint(0, 255, shape) for shape in shapes]
```
Comment (sandeep-krishnamurthy) [inline tests/keras/utils/np_utils_test.py]: negative test cases?
Ex: Invalid data_format, Invalid Tensor dimension etc

File: keras/backend/mxnet_backend.py (hunk 1 above)
Comment (karan6181) [reply keras/backend/mxnet_backend.py]: Yes, modified the code and now the warnings is shifted under `_convnd()` method. This results in having one warning message per operator. Thank-You.

File: keras/utils/np_utils.py (hunk 2 above)
Comment (karan6181) [reply keras/utils/np_utils.py]: - Initially, I thought it's just for MXNet backend and I have coded that way. But your point makes more sense to have it generic and not bind to anyone backends. I have modified the code and now it's more generic across backends.
- Absolutely. Handled the case where a user can now provide either a Numpy tensor or an array of Numpy tensor.
- Thank-You.

File: keras/utils/np_utils.py (hunk 3 above)
Comment (karan6181) [reply keras/utils/np_utils.py]: Modified. Thank-You.

File: keras/utils/np_utils.py (hunk 4 above)
Comment (karan6181) [reply keras/utils/np_utils.py]: Modified. Thank-You.

File: keras/utils/np_utils.py (hunk 5 above)
Comment (karan6181) [reply keras/utils/np_utils.py]: Done with help of `to_channels_first()` and `to_channels_first_helper()`

File: keras/utils/np_utils.py (hunk 6 above)
Comment (karan6181) [reply keras/utils/np_utils.py]: The method was written in such a way that if user calls `to_channels_first()` two or more times accidently on the same data tensor, then the method should return the ouput in always `channels_first` format. If I remove that check, then there is a channel conversion mismatch if user calls `to_channels_first()` multiple times.

```python
For instance:

Before check:
x_input = A Numpy tensor of shape (10, 3, 32, 32)
x_input = to_channels_first(x_input) -> output shape (10, 32, 3, 32)

After check:
x_input = A Numpy tensor of shape (10, 3, 32, 32)
x_input = to_channels_first(x_input) -> output shape (10, 3, 32, 32)
```

So, by keeping that shape check, we won't face such issues. However, as u mentioned, it will fail when image dimension is `(2, 2)` or `(1, 1)` or values smaller than channels in general. This would be possible if the data is synthetic and also, there is no other way to check based on tensors dimensions and impossible to cover all the edge cases. So, I have removed the check and make it more generic and assume user aware of it's data dimensions.

Thank-You.

File: tests/keras/utils/np_utils_test.py (hunk 7 above)
Comment (karan6181) [reply tests/keras/utils/np_utils_test.py]: Already handled Invalid Tensor dimension use cases where it throws a ValueError if the tensor dimension is incorrect.

Can you please elaborate what do mean by `invalid_data_format` means. Are your referring to having a use case to check `data_format` as `channels_first` or `channels_last`?

Thank-You.

File: keras/utils/np_utils.py [hunk 8]
```diff
@@ -48,3 +50,71 @@ def normalize(x, axis=-1, order=2):
     l2 = np.atleast_1d(np.linalg.norm(x, order, axis))
     l2[l2 == 0] = 1
     return x / np.expand_dims(l2, axis)
+
+def to_channels_first_helper(np_data):
+    """
+    Helper function to Transform the image tensor data from `channels_last`
+    format to `channels_first` format
+
+    # Arguments
```
Comment (sandeep-krishnamurthy) [inline keras/utils/np_utils.py]: Can be static?

File: keras/utils/np_utils.py [hunk 9]
```diff
@@ -48,3 +50,71 @@ def normalize(x, axis=-1, order=2):
...
+    Transform the image data to `channels_first` format.
+
+    # Arguments
+        data: A Numpy data tensor or a list of Numpy data tensor
+
+    # Returns
+        A Numpy data tensor or a list of Numpy data tensor with `channels_first`
+        format
+    """
+    if isinstance(data, list):
```
Comment (sandeep-krishnamurthy) [inline keras/utils/np_utils.py]: Check for None

Comment (karan6181) [reply keras/utils/np_utils.py]: Done. Checked for None as well as an empty list. Thank-You.
